bundles/alpha.model/src/alpha/model/util/Facet.xtend

```diff
@@ -77,18 +86,82 @@ class Facet {
 	
 	/** Extract the information from the given set. */
 	new(ISLBasicSet basicSet, ArrayList<Integer> saturatedInequalityIndices) {
-		equalities = DomainOperations.toISLEqualityMatrix(basicSet)
-		indexCount = basicSet.dim(ISLDimType.isl_dim_set)
-		indexInequalities = getInequalities(basicSet, indexCount, true)
+		val basicSetNoRedundancies = basicSet.copy.removeRedundancies
+		
+		space = basicSetNoRedundancies.space 
+		equalities = DomainOperations.toISLEqualityMatrix(basicSetNoRedundancies)
+		indexCount = basicSetNoRedundancies.dim(ISLDimType.isl_dim_set)
+		
+		val allInequalities = getInequalities(basicSetNoRedundancies, indexCount, true)
+		val inequalities = separateEffectivelySaturatedInequalities(allInequalities, space)
+		indexInequalities = inequalities.key
+		effectivelySaturatedInequalities = inequalities.value
+		
+		if (effectivelySaturatedInequalities.nbRows % 2 != 0) {
+			throw new Exception('Failed to create lattice, there should be an even number of (or zero) effectively saturated constraints')
+		}
+		
 		indexInequalityCount = indexInequalities.nbRows
-		isBounded = basicSet.bounded
-		isEmpty = basicSet.empty
+		isBounded = basicSetNoRedundancies.bounded
+		isEmpty = basicSetNoRedundancies.empty
 		parameterEqualityCount = countParameterConstraints(equalities, indexCount)
-		parameterInequalities = getInequalities(basicSet, indexCount, false)
+		parameterInequalities = getInequalities(basicSetNoRedundancies, indexCount, false)
 		this.saturatedInequalityIndices = saturatedInequalityIndices
-		space = basicSet.space 
+		
+		dimensionality = dimensionality(effectivelySaturatedInequalities, equalities, indexCount)
+	}
+	
+	/** Separates the effectively saturated inequalities from true unsaturated inequalities */
+	def static separateEffectivelySaturatedInequalities(ISLMatrix allInequalities, ISLSpace space) {
+		val matrix = allInequalities.copy.dropCols(space.nbParams + space.nbIndices, 1).toLongMatrix
+
+		// Use a hash map to get the index of a particular vector in matrix
+		val idxMap = new HashMap<String, Long>()
+		(0..<matrix.length).forEach[i | idxMap.put(matrix.get(i).toString, Long.valueOf(i))]
+
+		
+		// Identify the indices of any parallel constraints. For each vector in matrix, check
+		// to see if its negation exists (i.e., is a key in idxMap). Label such vectors as 
+		// having a colinear partner.
+		val negMatrix = matrix.length > 0 ? matrix.scalarMultiplication(-1) : matrix
```

Comment:
Let's move the logic for the `negMatrix` calculation into `MatrixOperations.scalarMultiplication` to avoid null reference exceptions in the future